File: core/src/main/java/com/shatteredpixel/shatteredpixeldungeon/levels/Terrain.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 *
 * Shattered Pixel Dungeon
 * Copyright (C) 2014-2019 Evan Debenham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.shatteredpixel.shatteredpixeldungeon.levels;

public class Terrain {

	public static final int CHASM			= 0;
	public static final int EMPTY			= 1;
	public static final int GRASS			= 2;
	public static final int EMPTY_WELL		= 3;
	public static final int WALL			= 4;
	public static final int DOOR			= 5;
	public static final int OPEN_DOOR		= 6;
	public static final int ENTRANCE		= 7;
	public static final int EXIT			= 8;
	public static final int EMBERS			= 9;
	public static final int LOCKED_DOOR		= 10;
	public static final int PEDESTAL		= 11;
	public static final int WALL_DECO		= 12;
	public static final int BARRICADE		= 13;
	public static final int EMPTY_SP		= 14;
	public static final int HIGH_GRASS		= 15;
	public static final int FURROWED_GRASS	= 30;

	public static final int SECRET_DOOR	    = 16;
	public static final int SECRET_TRAP     = 17;
	public static final int TRAP            = 18;
	public static final int INACTIVE_TRAP   = 19;

	public static final int EMPTY_DECO		= 20;
	public static final int LOCKED_EXIT		= 21;
	public static final int UNLOCKED_EXIT	= 22;
	public static final int SIGN			= 23;
	public static final int WELL			= 24;
	public static final int STATUE			= 25;
	public static final int STATUE_SP		= 26;
	public static final int BOOKSHELF		= 27;
	public static final int ALCHEMY			= 28;

	public static final int WATER		    = 29;
	
	public static final int PASSABLE		= 0x01;
	public static final int LOS_BLOCKING	= 0x02;
	public static final int FLAMABLE		= 0x04;
	public static final int SECRET			= 0x08;
	public static final int SOLID			= 0x10;
	public static final int AVOID			= 0x20;
	public static final int LIQUID			= 0x40;
	public static final int PIT				= 0x80;
	
	public static final int[] flags = new int[256];
	static {
		flags[CHASM]		= AVOID	| PIT;
		flags[EMPTY]		= PASSABLE;
		flags[GRASS]		= PASSABLE | FLAMABLE;
		flags[EMPTY_WELL]	= PASSABLE;
		flags[WATER]		= PASSABLE | LIQUID;
		flags[WALL]			= LOS_BLOCKING | SOLID;
		flags[DOOR]			= PASSABLE | LOS_BLOCKING | FLAMABLE | SOLID;
		flags[OPEN_DOOR]	= PASSABLE | FLAMABLE;
		flags[ENTRANCE]		= PASSABLE/* | SOLID*/;
		flags[EXIT]			= PASSABLE;
		flags[EMBERS]		= PASSABLE;
		flags[LOCKED_DOOR]	= LOS_BLOCKING | SOLID;
		flags[PEDESTAL]		= PASSABLE;
		flags[WALL_DECO]	= flags[WALL];
		flags[BARRICADE]	= FLAMABLE | SOLID | LOS_BLOCKING;
		flags[EMPTY_SP]		= flags[EMPTY];
		flags[HIGH_GRASS]	= PASSABLE | LOS_BLOCKING | FLAMABLE;
		flags[FURROWED_GRASS]= flags[HIGH_GRASS];

		flags[SECRET_DOOR]  = flags[WALL]  | SECRET;
		flags[SECRET_TRAP]  = flags[EMPTY] | SECRET;
		flags[TRAP]         = AVOID;
		flags[INACTIVE_TRAP]= flags[EMPTY];

		flags[EMPTY_DECO]	= flags[EMPTY];
		flags[LOCKED_EXIT]	= SOLID;
		flags[UNLOCKED_EXIT]= PASSABLE;
		flags[SIGN]			= SOLID; //Currently these are unused except for visual tile overrides where we want terrain to be solid with no other properties
		flags[WELL]			= AVOID;
		flags[STATUE]		= SOLID;
		flags[STATUE_SP]	= flags[STATUE];
		flags[BOOKSHELF]	= flags[BARRICADE];
		flags[ALCHEMY]		= SOLID;

	}

	public static int discover( int terr ) {
		switch (terr) {
		case SECRET_DOOR:
			return DOOR;
		case SECRET_TRAP:
			return TRAP;
		default:
			return terr;
		}
	}

	//removes signs, places floors instead
	public static int[] convertTilesFrom0_6_0b(int[] map){
		for (int i = 0; i < map.length; i++){
			if (map[i] == 23){
				map[i] = 1;
			}
		}
		return map;
	}

}

File: pixel-dungeon-web/src/levels/Terrain.js
/*
 * Pixel Dungeon Web Port
 * Ported from Shattered Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 * Copyright (C) 2014-2025 Evan Debenham
 *
 * Terrain Types and Flags
 * Source: core/src/main/java/com/shatteredpixel/shatteredpixeldungeon/levels/Terrain.java
 */

/**
 * Terrain tile types
 * Source: Terrain.java:26-68
 */
export const TerrainType = {
    CHASM: 0,
    EMPTY: 1,
    GRASS: 2,
    EMPTY_WELL: 3,
    WALL: 4,
    DOOR: 5,
    OPEN_DOOR: 6,
    ENTRANCE: 7,
    ENTRANCE_SP: 37,
    EXIT: 8,
    EMBERS: 9,
    LOCKED_DOOR: 10,
    CRYSTAL_DOOR: 31,
    PEDESTAL: 11,
    WALL_DECO: 12,
    BARRICADE: 13,
    EMPTY_SP: 14,
    HIGH_GRASS: 15,
    FURROWED_GRASS: 30,

    SECRET_DOOR: 16,
    SECRET_TRAP: 17,
    TRAP: 18,
    INACTIVE_TRAP: 19,

    EMPTY_DECO: 20,
    LOCKED_EXIT: 21,
    UNLOCKED_EXIT: 22,
    WELL: 24,
    BOOKSHELF: 27,
    ALCHEMY: 28,

    CUSTOM_DECO_EMPTY: 32,
    CUSTOM_DECO: 23,
    STATUE: 25,
    STATUE_SP: 26,

    REGION_DECO: 33,
    REGION_DECO_ALT: 34,
    MINE_CRYSTAL: 35,
    MINE_BOULDER: 36,

    WATER: 29
};

/**
 * Terrain property flags
 * Source: Terrain.java:71-78
 */
export const TerrainFlags = {
    PASSABLE: 0x01,
    LOS_BLOCKING: 0x02,
    FLAMABLE: 0x04,
    SECRET: 0x08,
    SOLID: 0x10,
    AVOID: 0x20,
    LIQUID: 0x40,
    PIT: 0x80
};

/**
 * Terrain flags lookup table
 * Each terrain type maps to a combination of flags
 * Source: Terrain.java:80-125
 */
export const flags = new Array(256).fill(0);

// Initialize flags for each terrain type
// Source: Terrain.java:82-124
flags[TerrainType.CHASM] = TerrainFlags.AVOID | TerrainFlags.PIT;
flags[TerrainType.EMPTY] = TerrainFlags.PASSABLE;
flags[TerrainType.GRASS] = TerrainFlags.PASSABLE | TerrainFlags.FLAMABLE;
flags[TerrainType.EMPTY_WELL] = TerrainFlags.PASSABLE;
flags[TerrainType.WATER] = TerrainFlags.PASSABLE | TerrainFlags.LIQUID;
flags[TerrainType.WALL] = TerrainFlags.LOS_BLOCKING | TerrainFlags.SOLID;
flags[TerrainType.DOOR] = TerrainFlags.PASSABLE | TerrainFlags.LOS_BLOCKING | TerrainFlags.FLAMABLE | TerrainFlags.SOLID;
flags[TerrainType.OPEN_DOOR] = TerrainFlags.PASSABLE | TerrainFlags.FLAMABLE;
flags[TerrainType.ENTRANCE] = TerrainFlags.PASSABLE;
flags[TerrainType.ENTRANCE_SP] = flags[TerrainType.ENTRANCE];
flags[TerrainType.EXIT] = TerrainFlags.PASSABLE;
flags[TerrainType.EMBERS] = TerrainFlags.PASSABLE;
flags[TerrainType.LOCKED_DOOR] = TerrainFlags.LOS_BLOCKING | TerrainFlags.SOLID;
flags[TerrainType.CRYSTAL_DOOR] = TerrainFlags.SOLID;
flags[TerrainType.PEDESTAL] = TerrainFlags.PASSABLE;
flags[TerrainType.WALL_DECO] = flags[TerrainType.WALL];
flags[TerrainType.BARRICADE] = TerrainFlags.FLAMABLE | TerrainFlags.SOLID | TerrainFlags.LOS_BLOCKING;
flags[TerrainType.EMPTY_SP] = flags[TerrainType.EMPTY];
flags[TerrainType.HIGH_GRASS] = TerrainFlags.PASSABLE | TerrainFlags.LOS_BLOCKING | TerrainFlags.FLAMABLE;
flags[TerrainType.FURROWED_GRASS] = flags[TerrainType.HIGH_GRASS];

flags[TerrainType.SECRET_DOOR] = flags[TerrainType.WALL] | TerrainFlags.SECRET;
flags[TerrainType.SECRET_TRAP] = flags[TerrainType.EMPTY] | TerrainFlags.SECRET;
flags[TerrainType.TRAP] = TerrainFlags.AVOID;
flags[TerrainType.INACTIVE_TRAP] = flags[TerrainType.EMPTY];

flags[TerrainType.EMPTY_DECO] = flags[TerrainType.EMPTY];
flags[TerrainType.LOCKED_EXIT] = TerrainFlags.SOLID;
flags[TerrainType.UNLOCKED_EXIT] = TerrainFlags.PASSABLE;
flags[TerrainType.WELL] = TerrainFlags.AVOID;
flags[TerrainType.BOOKSHELF] = flags[TerrainType.BARRICADE];
flags[TerrainType.ALCHEMY] = TerrainFlags.SOLID;

flags[TerrainType.CUSTOM_DECO_EMPTY] = flags[TerrainType.EMPTY];
flags[TerrainType.CUSTOM_DECO] = TerrainFlags.SOLID;
flags[TerrainType.STATUE] = TerrainFlags.SOLID;
flags[TerrainType.STATUE_SP] = flags[TerrainType.STATUE];

flags[TerrainType.REGION_DECO] = flags[TerrainType.STATUE];
flags[TerrainType.REGION_DECO_ALT] = flags[TerrainType.STATUE_SP];
flags[TerrainType.MINE_CRYSTAL] = TerrainFlags.SOLID;
flags[TerrainType.MINE_BOULDER] = TerrainFlags.SOLID;

/**
 * Discover hidden terrain (reveal secrets)
 * Source: Terrain.java:127-136
 */
export function discover(terr) {
    switch (terr) {
        case TerrainType.SECRET_DOOR:
            return TerrainType.DOOR;
        case TerrainType.SECRET_TRAP:
            return TerrainType.TRAP;
        default:
            return terr;
    }
}

/**
 * Check if terrain has a specific flag
 */
export function hasFlag(terr, flag) {
    return (flags[terr] & flag) !== 0;
}

/**
 * Check if terrain is passable
 */
export function isPassable(terr) {
    return hasFlag(terr, TerrainFlags.PASSABLE);
}

/**
 * Check if terrain blocks line of sight
 */
export function losBlocking(terr) {
    return hasFlag(terr, TerrainFlags.LOS_BLOCKING);
}

/**
 * Check if terrain is solid
 */
export function isSolid(terr) {
    return hasFlag(terr, TerrainFlags.SOLID);
}

/**
 * Check if terrain should be avoided
 */
export function isAvoid(terr) {
    return hasFlag(terr, TerrainFlags.AVOID);
}
